File: src/python/tests/test_calculator.py
"""
Unit tests for the calculator module.

These tests demonstrate:
- Basic test structure
- Testing different scenarios
- Testing error conditions
- Parametrized tests
"""

import pytest
from calculator import add, subtract, multiply, divide


class TestAdd:
    """Tests for the add function."""

    def test_add_positive_numbers(self):
        """Test adding two positive numbers."""
        assert add(2, 3) == 5
        assert add(10, 20) == 30

    def test_add_negative_numbers(self):
        """Test adding negative numbers."""
        assert add(-5, -3) == -8
        assert add(-10, 10) == 0

    def test_add_zero(self):
        """Test adding zero."""
        assert add(0, 0) == 0
        assert add(5, 0) == 5
        assert add(0, 5) == 5

    @pytest.mark.parametrize("a,b,expected", [
        (1, 1, 2),
        (100, 200, 300),
        (-50, 50, 0),
        (0, 0, 0),
    ])
    def test_add_parametrized(self, a, b, expected):
        """Test add with multiple input combinations."""
        assert add(a, b) == expected


class TestSubtract:
    """Tests for the subtract function."""

    def test_subtract_positive_numbers(self):
        """Test subtracting positive numbers."""
        assert subtract(5, 3) == 2
        assert subtract(10, 5) == 5

    def test_subtract_negative_numbers(self):
        """Test subtracting negative numbers."""
        assert subtract(-5, -3) == -2
        assert subtract(5, -3) == 8

    def test_subtract_zero(self):
        """Test subtracting zero."""
        assert subtract(5, 0) == 5
        assert subtract(0, 5) == -5
        assert subtract(0, 0) == 0

    def test_subtract_larger_from_smaller(self):
        """Test subtracting larger number from smaller."""
        assert subtract(3, 5) == -2
        assert subtract(10, 20) == -10


class TestMultiply:
    """Tests for the multiply function."""

    def test_multiply_positive_numbers(self):
        """Test multiplying positive numbers."""
        assert multiply(2, 3) == 6
        assert multiply(5, 4) == 20

    def test_multiply_negative_numbers(self):
        """Test multiplying with negative numbers."""
        assert multiply(-2, 3) == -6
        assert multiply(-2, -3) == 6

    def test_multiply_by_zero(self):
        """Test multiplying by zero."""
        assert multiply(5, 0) == 0
        assert multiply(0, 5) == 0
        assert multiply(0, 0) == 0

    def test_multiply_by_one(self):
        """Test multiplying by one."""
        assert multiply(5, 1) == 5
        assert multiply(1, 5) == 5


class TestDivide:
    """Tests for the divide function."""

    def test_divide_positive_numbers(self):
        """Test dividing positive numbers."""
        assert divide(10, 2) == 5.0
        assert divide(15, 3) == 5.0

    def test_divide_with_float_result(self):
        """Test division with float result."""
        assert divide(7, 2) == 3.5
        assert divide(5, 4) == 1.25

    def test_divide_negative_numbers(self):
        """Test dividing with negative numbers."""
        assert divide(-10, 2) == -5.0
        assert divide(10, -2) == -5.0
        assert divide(-10, -2) == 5.0

    def test_divide_by_zero_raises_error(self):
        """Test that dividing by zero raises ZeroDivisionError."""
        with pytest.raises(ZeroDivisionError):
            divide(10, 0)

    def test_divide_zero_by_number(self):
        """Test dividing zero by a number."""
        assert divide(0, 5) == 0.0
        assert divide(0, -5) == 0.0

File: src/c/tests/test_calculator.c
/*
 * Unit tests for calculator functions.
 *
 * This demonstrates a simple testing framework for C code.
 */

#include <stdio.h>
#include <math.h>
#include "../calculator.h"

/* Test statistics */
static int tests_run = 0;
static int tests_passed = 0;
static int tests_failed = 0;

/* Color codes for output */
#define COLOR_GREEN "\x1b[32m"
#define COLOR_RED "\x1b[31m"
#define COLOR_RESET "\x1b[0m"

/* Test helper macros */
#define TEST(name) static void name(void)

#define RUN_TEST(test) do { \
	printf("Running %s... ", #test); \
	tests_run++; \
	test(); \
} while(0)

#define ASSERT_EQUAL(a, b) do { \
	if ((a) != (b)) { \
		printf(COLOR_RED "FAILED" COLOR_RESET "\n"); \
		printf("  Expected: %d, Got: %d\n", (int)(b), (int)(a)); \
		printf("  File: %s, Line: %d\n", __FILE__, __LINE__); \
		tests_failed++; \
		return; \
	} \
} while(0)

#define ASSERT_FLOAT_EQUAL(a, b) do { \
	if (fabs((a) - (b)) > 0.0001) { \
		printf(COLOR_RED "FAILED" COLOR_RESET "\n"); \
		printf("  Expected: %f, Got: %f\n", (double)(b), (double)(a)); \
		printf("  File: %s, Line: %d\n", __FILE__, __LINE__); \
		tests_failed++; \
		return; \
	} \
} while(0)

#define PASS() do { \
	printf(COLOR_GREEN "PASSED" COLOR_RESET "\n"); \
	tests_passed++; \
} while(0)

/* Test cases for add function */
TEST(test_add_positive_numbers)
{
	ASSERT_EQUAL(add(2, 3), 5);
	ASSERT_EQUAL(add(10, 20), 30);
	PASS();
}

TEST(test_add_negative_numbers)
{
	ASSERT_EQUAL(add(-5, -3), -8);
	ASSERT_EQUAL(add(-10, 10), 0);
	PASS();
}

TEST(test_add_zero)
{
	ASSERT_EQUAL(add(0, 0), 0);
	ASSERT_EQUAL(add(5, 0), 5);
	ASSERT_EQUAL(add(0, 5), 5);
	PASS();
}

/* Test cases for subtract function */
TEST(test_subtract_positive_numbers)
{
	ASSERT_EQUAL(subtract(5, 3), 2);
	ASSERT_EQUAL(subtract(10, 5), 5);
	PASS();
}

TEST(test_subtract_negative_numbers)
{
	ASSERT_EQUAL(subtract(-5, -3), -2);
	ASSERT_EQUAL(subtract(5, -3), 8);
	PASS();
}

TEST(test_subtract_zero)
{
	ASSERT_EQUAL(subtract(5, 0), 5);
	ASSERT_EQUAL(subtract(0, 5), -5);
	ASSERT_EQUAL(subtract(0, 0), 0);
	PASS();
}

/* Test cases for multiply function */
TEST(test_multiply_positive_numbers)
{
	ASSERT_EQUAL(multiply(2, 3), 6);
	ASSERT_EQUAL(multiply(5, 4), 20);
	PASS();
}

TEST(test_multiply_negative_numbers)
{
	ASSERT_EQUAL(multiply(-2, 3), -6);
	ASSERT_EQUAL(multiply(-2, -3), 6);
	PASS();
}

TEST(test_multiply_by_zero)
{
	ASSERT_EQUAL(multiply(5, 0), 0);
	ASSERT_EQUAL(multiply(0, 5), 0);
	ASSERT_EQUAL(multiply(0, 0), 0);
	PASS();
}

/* Test cases for divide function */
TEST(test_divide_positive_numbers)
{
	ASSERT_FLOAT_EQUAL(divide(10.0, 2.0), 5.0);
	ASSERT_FLOAT_EQUAL(divide(15.0, 3.0), 5.0);
	PASS();
}

TEST(test_divide_with_float_result)
{
	ASSERT_FLOAT_EQUAL(divide(7.0, 2.0), 3.5);
	ASSERT_FLOAT_EQUAL(divide(5.0, 4.0), 1.25);
	PASS();
}

TEST(test_divide_negative_numbers)
{
	ASSERT_FLOAT_EQUAL(divide(-10.0, 2.0), -5.0);
	ASSERT_FLOAT_EQUAL(divide(10.0, -2.0), -5.0);
	ASSERT_FLOAT_EQUAL(divide(-10.0, -2.0), 5.0);
	PASS();
}

TEST(test_divide_by_zero)
{
	/* divide by zero returns 0.0 and prints error */
	ASSERT_FLOAT_EQUAL(divide(10.0, 0.0), 0.0);
	PASS();
}

TEST(test_divide_zero_by_number)
{
	ASSERT_FLOAT_EQUAL(divide(0.0, 5.0), 0.0);
	ASSERT_FLOAT_EQUAL(divide(0.0, -5.0), 0.0);
	PASS();
}

/* Main test runner */
int main(void)
{
	printf("\n");
	printf("========================================\n");
	printf("   Calculator Unit Tests\n");
	printf("========================================\n\n");

	/* Run add tests */
	printf("Testing add function:\n");
	RUN_TEST(test_add_positive_numbers);
	RUN_TEST(test_add_negative_numbers);
	RUN_TEST(test_add_zero);

	/* Run subtract tests */
	printf("\nTesting subtract function:\n");
	RUN_TEST(test_subtract_positive_numbers);
	RUN_TEST(test_subtract_negative_numbers);
	RUN_TEST(test_subtract_zero);

	/* Run multiply tests */
	printf("\nTesting multiply function:\n");
	RUN_TEST(test_multiply_positive_numbers);
	RUN_TEST(test_multiply_negative_numbers);
	RUN_TEST(test_multiply_by_zero);

	/* Run divide tests */
	printf("\nTesting divide function:\n");
	RUN_TEST(test_divide_positive_numbers);
	RUN_TEST(test_divide_with_float_result);
	RUN_TEST(test_divide_negative_numbers);
	RUN_TEST(test_divide_by_zero);
	RUN_TEST(test_divide_zero_by_number);

	/* Print summary */
	printf("\n========================================\n");
	printf("Test Results:\n");
	printf("  Total:  %d\n", tests_run);
	printf("  " COLOR_GREEN "Passed: %d" COLOR_RESET "\n", tests_passed);
	if (tests_failed > 0) {
		printf("  " COLOR_RED "Failed: %d" COLOR_RESET "\n", tests_failed);
	}
	printf("========================================\n\n");

	if (tests_failed == 0 && tests_run == tests_passed) {
		printf(COLOR_GREEN "✓ All tests passed!" COLOR_RESET "\n\n");
		return 0;
	} else {
		printf(COLOR_RED "✗ Some tests failed!" COLOR_RESET "\n\n");
		return 1;
	}
}
